### DISCUSSION TITLE
Question: How to use different http message handlers for CallHttpAsync()

### DISCUSSION THREAD

**Discussion by riccardobecker** [Azure/azure-functions-durable-extension#1763] [Q&A] [answered]
Hi,

currently i am building durable functions and need to reach out to different endpoints from the orchestrator. I have created a custom implementation of IDurableHttpMessageHandlerFactory and added client certificate from keyvault because for one external endpoint i need to authenticate like that. 

For a different endpoint, i dont need this and i dont want to use the same handler for this.

In startup i have:

builder.Services.AddSingleton<IDurableHttpMessageHandlerFactory, DurableHttpHandlerFactory>();

but i dont think there is another of adding multiple implementations of IDurableHttpMessageHandlerFactory?

What would be a proper way to support different httpclients from durables?

thanks!

**Comment by ConnorMcMahon** [comment, marked as answer]
We do not have a way to have different HttpMessageHandlers for different calls to `CallHttpAsync()`.

We could have theoretically supported this if we had an optional parameter on `IDurableHttpMessageHandlerFactory.CreateHttpMessageHandler()` to specify a handler name, and let `CallHttpAsync()` pass in handler names, but since the interface is public, we can't add new optional parameters in a non-breaking way.

The way to work around this in the meantime would be to return a HttpMessageHandler that is itself a composite `HttpMessageHandler`. It can observe characteristics of the HTTP request (such as the endpoint being targeted, headers on the request, etc.) and make a decision about which of it's internal `HttpMessageHandler` implementations to use. 

For example:

```csharp
public class CertificateHandler : DelegatingHandler
{
    protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        if (IsCertAuthRequest(request))
        {
            // Logic to add cert based authentication to request
        }

        return base.SendAsync(request, cancellationToken);
    }

    private static bool IsCertAuthRequest(HttpRequestMessage request)
    {
        // Look at HTTP request to decide if we should handle it in this layer
        return true;
    }
}
```

You would then instantiate that object like

```csharp
new CertificateHandler(new HttpClientHandler())
```

**Comment by riccardobecker** [comment]
Connor, thanks for your feedback. And for the direction on how to handle this. This approach seems viable yes.

regards
